File: 035-search-insert-position/search-insert-position.py
# -*- coding:utf-8 -*-


# Given a sorted array and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.
#
# You may assume no duplicates in the array.
#
#
# Here are few examples.
# [1,3,5,6], 5 &#8594; 2
# [1,3,5,6], 2 &#8594; 1
# [1,3,5,6], 7 &#8594; 4
# [1,3,5,6], 0 &#8594; 0


class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums:
            return 0
        
        for idx,val in enumerate(nums):
            if val >= target:
                return idx
        
        return len(nums)

File: solutions/035-search-insert-position/search-insert-position.cpp
// Given a sorted array and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.
//
// You may assume no duplicates in the array.
//
// Example 1:
//
//
// Input: [1,3,5,6], 5
// Output: 2
//
//
// Example 2:
//
//
// Input: [1,3,5,6], 2
// Output: 1
//
//
// Example 3:
//
//
// Input: [1,3,5,6], 7
// Output: 4
//
//
// Example 4:
//
//
// Input: [1,3,5,6], 0
// Output: 0
//
//


class Solution {
public:
    int searchInsert(vector<int>& nums, int t) {
        int l = 0;
        int r = nums.size() - 1;
        while (l <= r) {
            int m = (l + r) >> 1;
            if (nums[m] < t) {
                l = m + 1;
            } else if (nums[m] > t) {
                r = m - 1;
            } else {
                return m;
            }
        }
        return l;
    }
};
